activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -22,6 +22,16 @@ def method_missing(*)
       end
       NULL_CONFIG = NullConfig.new # :nodoc:
 
+      def initialize
+        super()
+        @mutex = Mutex.new
+        @server_version = nil
+      end
+
+      def server_version(connection) # :nodoc:
+        @server_version || @mutex.synchronize { @server_version ||= connection.get_database_version }
```

Comment:
I'm tempted to rename `Adapter#database_version` into `#server_version` as IMO it's less ambiguous, but that would break third party adapters, so might not be worth it.